### ISSUE TITLE
SegmentAnything Model, memory footprint way bigger in C# compared to Python 

### ISSUE THREAD

**Issue by LamerTex** [dotnet/TorchSharp#1002]
I'm trying to port the [SAM model from the original model in Python](https://github.com/facebookresearch/segment-anything)  to C# using this wonderful library.

After a bit of trouble, I was able to translate the model to C#, load the weights by converting with the script the original ".pth" files to the correct format for TorchSharp, and finally make it run and obtain the same results as the original model in both CPU and CUDA devices.

The problem is that in CUDA I'm only able to do this with the "smaller" sam_vit_b_01ec64.pth and not the full "sam_vit_h_4b8939.pth" because the memory footprint of the C# version is much bigger compared to the python (almost 12GB vs 4GB out of 12GB available in my GPU). 

I've read several time the MemoryManagement wiki page, and I'm already implementing [the third technique](https://github.com/dotnet/TorchSharp/wiki/Memory-Management#technique-3-torchnewdisposescope) for the whole model at the best of my capacity, but comparing "line by line" the two models, there seems to be some operation which in Python seems "free" while in C# occupy more memory.

I'll show you one example of this, but I've found several others which together results in the 12GB VRAM vs the 4GB for python:

This is the preprocess function in sam.py, the operation that has the problem simply does a subtraction and a division between a tensor "x" of tpye uint8 with shape [1, 3, 683, 1024] and 2 tensor (self.pixel_mean and self.pixel_std) of type float32 with shape [3,1,1].

  ```
  def preprocess(self, x: torch.Tensor) -> torch.Tensor:
          """Normalize pixel values and pad to a square input."""
          # Normalize colors
          x = (x - self.pixel_mean) / self.pixel_std
  
          # Pad
          h, w = x.shape[-2:]
          padh = self.image_encoder.img_size - h
          padw = self.image_encoder.img_size - w
          x = F.pad(x, (0, padw, 0, padh))
          return x
  ```
  
  this operation in Python doesn't grow the GPU memory at all, it seems like it is reusing the same memory and overwriting it with the new values? But the operation it's also changing the type of "x" from uint8 to float32 so I don't know how it does it.
  While in C# it has been translated with the method:
  ```
  public Tensor preprocess(Tensor x)
  {
      using (DisposeScope disposer = torch.NewDisposeScope())
      {
          //    """Normalize pixel values and pad to a square input."""
          //# Normalize colors
          x = (x - get_buffer("pixel_mean")) / get_buffer("pixel_std");
  
          //# Pad
          long[] hw = x.shape;
          int h = (int)hw[2];
          int w = (int)hw[3];
  
          int padh = image_encoder.img_size - h;
          int padw = image_encoder.img_size - w;
          x = nn.functional.pad(x, new long[] { 0, padw, 0, padh });
          return x.MoveToOuterDisposeScope();
      }
  }
  ```
this time the same operation increment the GPU memory by 400MB! 
I've also tried to "decompose" the operation into several smaller ones so it becomes 
```
x1 = x -  self.pixel_mean
x2 = x1 /  self.pixel_std
```
and similar in C# but the result is the same, 0 increment in Python and 400MB more in C# at the first subtraction (no increment for the division). 
I've also tried using the explicit operators torch.sub/torch.div and also starting already with a Float32 "x" instead of an Int8 one without any difference.

What I'm missing here? 
I'm doing something wrong or is the original pytorch in Python doing some magic that cannot be replicated by TorchSharp? 

In CPU the problem remains, but seems different. 
The operation I've used here to show the difference in memory for GPU doesn't seem to behave different from Python to C#, but at the end of the encoding process the memory in Python remains the same (just a single tensor should remain in memory), in C# instead 7GB of RAM remain occupied and they are freed only after disposing the last Tensor with the image encoding. 
It seems like the last tensor is "keeping" all the others from being actually disposed, even thought the "dispose scope" is used everywhere. 

Thanks for your time!

**Comment by LamerTex** [comment]
Just after finishing writing this issue, I've investigated more on the CPU side, in particular, the "It seems like the last tensor is "keeping" all the others from being actually disposed, even though the "dispose scope" is used everywhere."

This was the key to finding the problem -> the last tensor is keeping all the others.

As soon as the last tensor was disposed all the memory was finally released, so since I'm a noob in pytorch I've tried to simply "clone" the last tensor and dispose the old one. 
The clone method wasn't working so I tried with a "detach" and this time it worked, all the memory was free and I still had a tensor with the image encoding I needed. 

Of course, doing this only at the end for the last tensor would still fill the memory before releasing it all together.
Instead, I've simply searched for all calls to ".MoveToOuterDisposeScope()" and changed them with ".detach().MoveToOuterDisposeScope()" and now the memory in CPU is the same, and also with CUDA is now similar (4.2GB in python vs 4.9GB in C# for the B model) and I can finally use also the "best model -> sam_vit_h_4b8939.pth" without finishing the VRAM ( 8.1GB in C#) and with similar time performance to python and pytorch.


I'm pretty sure that calling "detach()" is not the correct thing to do, otherwise I'm pretty sure the "DisposeScope" would already call it, or at least it would have been nominated in the MemoryManagement wiki page. 
It is more likely that I'm doing something very wrong, do you know what could it be?

**Comment by GeorgeS2019** [comment]
@LamerTex  I am not in any position to help you in any technical way. I just want to thank you for trying TorchSharp on such ambitious project. It will be a greet boost of morale to TorchSharp community.

**Comment by NiklasGustafsson** [comment]
The tensors are being kept around because of their gradients are thought to be needed later, I'm pretty sure. If you can call `detach()` on the result, I'm guessing that this logic isn't really part of the model itself (as 'preprocessing' also suggests)

I wonder if a `using (var ng = torch.no_grad())` at the top of the function would make a difference.

**Comment by NiklasGustafsson** [comment]
BTW, for this:
```C#
x = nn.functional.pad(x, new long[] { 0, padw, 0, padh });
```

There's an overload of `pad()` that takes a quadruple for the padding, so you shouldn't have to create the array.

**Comment by LamerTex** [comment]
> The tensors are being kept around because their gradients are thought to be needed later, I'm pretty sure. Since you can call `detach()` on the result without problems, I'm guessing that this logic isn't really part of the model itself (as the name 'preprocess' also suggests)
> 
> I wonder if a `using (var ng = torch.no_grad())` at the top of the function, or at the call site would make a difference.

Yes this was it, thank you! 
The operation is the computation of the "encoding" of the input image and it is all inference and not training (since the model has been trained already by Facebook :D )

It was also just a mistake by me, the "no_grad" was already present in the python code, just it is a decoration of the method so I didn't see it at all 

```
@torch.no_grad()
    def set_torch_image(
        self,
        transformed_image: torch.Tensor,
        original_image_size: Tuple[int, ...],
    ) -> None:
        """
        Calculates the image embeddings for the provided image, allowing
        masks to be predicted with the 'predict' method. Expects the input
        image to be already transformed to the format expected by the model.

        Arguments:
          transformed_image (torch.Tensor): The input image, with shape
            1x3xHxW, which has been transformed with ResizeLongestSide.
          original_image_size (tuple(int, int)): The size of the image
            before transformation, in (H, W) format.
        """
        assert (
            len(transformed_image.shape) == 4
            and transformed_image.shape[1] == 3
            and max(*transformed_image.shape[2:]) == self.model.image_encoder.img_size
        ), f"set_torch_image input must be BCHW with long side {self.model.image_encoder.img_size}."
        self.reset_image()

        self.original_size = original_image_size
        self.input_size = tuple(transformed_image.shape[-2:])
        input_image = self.model.preprocess(transformed_image)
        self.features = self.model.image_encoder(input_image)
        self.is_image_set = True
```

With the no_grad() I was able to remove all the detach() calls and it works correctly.

Thanks again for this wonderful library!

**Comment by NiklasGustafsson** [comment]
Also, make sure to call `eval()` on the model before running it in inference! I believe that TorchScript will retain the `training` setting that was in place when it was exported from pytorch, so you want to make sure...
